### ISSUE TITLE
Fix COINBASE opcode impl

### ISSUE THREAD

**Issue by CPerezz** [appliedzkp/zkevm-circuits#289]
While working with @ed255 solving some Lookup issues related to the `CALLER` opcode impl, we started to add `RandomLinearCombination`s for the caller field as we need to match values from the `stack` and the `rw` tables.

One thing we noticed is that `COINBASE` opcode (implemented in https://github.com/appliedzkp/zkevm-circuits/pull/259)  was using the default block. And the default one uses `0x00..000` as addr for the coinbase. So we were curious to see if it was handling the RLC correctly (as the RLC of 0 is 0 too and that allows to match in all of the tables).

This is the part of the code where we did the changes:
https://github.com/appliedzkp/zkevm-circuits/blob/1c1a89b3f21d0d250cf15b69566b470466e99cef/bus-mapping/src/mock.rs#L29-L35

The run with `Coinbnase = 0` passes.
```
$ RUST_BACKTRACE=1 RUST_LOG=circuits,bus_mapping=trace cargo test coinbase_gadget_test -- --nocapture

running 1 test
load_rws: RwMap(
    {
        AccountDestructed: [],
        TxRefund: [],
        AccountStorage: [],
        TxAccessListAccount: [],
        Stack: [
            Stack {
                rw_counter: 1,
                is_write: true,
                call_id: 1,
                stack_pointer: 1023,
                value: 0,
            },
        ],
        TxAccessListAccountStorage: [],
        CallContext: [],
        Memory: [],
        Account: [],
    },
)
row: 0, col: Column { index: 4, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 5, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 6, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000008
row: 0, col: Column { index: 7, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 8, column_type: Advice }, val: 0x00000000000000000000000000000000000000000000000000000000000003ff
row: 0, col: Column { index: 9, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 10, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 11, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 4, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 5, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 6, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000008
row: 0, col: Column { index: 7, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 8, column_type: Advice }, val: 0x00000000000000000000000000000000000000000000000000000000000003ff
row: 0, col: Column { index: 9, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 10, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 11, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
test evm_circuit::execution::coinbase::test::coinbase_gadget_test has been running for over 60 seconds
test evm_circuit::execution::coinbase::test::coinbase_gadget_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 55 filtered out; finished in 62.80s
```

Once we edit coinbase field of `new_block` to be: `*COINBASE` (the lazy static).
```
$ RUST_BACKTRACE=1 RUST_LOG=circuits,bus_mapping=trace cargo test coinbase_gadget_test -- --nocapture

running 1 test
load_rws: RwMap(
    {
        Account: [],
        CallContext: [],
        TxAccessListAccountStorage: [],
        TxRefund: [],
        AccountStorage: [],
        TxAccessListAccount: [],
        AccountDestructed: [],
        Stack: [
            Stack {
                rw_counter: 1,
                is_write: true,
                call_id: 1,
                stack_pointer: 1023,
                value: 3222583902,
            },
        ],
        Memory: [],
    },
)
row: 0, col: Column { index: 4, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 5, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 6, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000008
row: 0, col: Column { index: 7, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 8, column_type: Advice }, val: 0x00000000000000000000000000000000000000000000000000000000000003ff
row: 0, col: Column { index: 9, column_type: Advice }, val: 0x2174a8cbf65a411a6fee0b3a86fb1cfa64cb765eb583bde86b5ebddbfcc43974
row: 0, col: Column { index: 10, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 11, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 4, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 5, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 6, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000008
row: 0, col: Column { index: 7, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000001
row: 0, col: Column { index: 8, column_type: Advice }, val: 0x00000000000000000000000000000000000000000000000000000000000003ff
row: 0, col: Column { index: 9, column_type: Advice }, val: 0x2174a8cbf65a411a6fee0b3a86fb1cfa64cb765eb583bde86b5ebddbfcc43974
row: 0, col: Column { index: 10, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
row: 0, col: Column { index: 11, column_type: Advice }, val: 0x0000000000000000000000000000000000000000000000000000000000000000
test evm_circuit::execution::coinbase::test::coinbase_gadget_test has been running for over 60 seconds
thread 'evm_circuit::execution::coinbase::test::coinbase_gadget_test' panicked at 'assertion failed: `(left == right)`
  left: `Err([Lookup { lookup_index: 154, row: 0 }])`,
 right: `Ok(())`', zkevm-circuits/src/evm_circuit/execution/coinbase.rs:105:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d914f17ca71a33a89b2dc3436fca51b1a091559e/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/d914f17ca71a33a89b2dc3436fca51b1a091559e/library/core/src/panicking.rs:106:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/d914f17ca71a33a89b2dc3436fca51b1a091559e/library/core/src/panicking.rs:144:5
   4: zkevm_circuits::evm_circuit::execution::coinbase::test::test_ok
             at ./src/evm_circuit/execution/coinbase.rs:105:9
   5: zkevm_circuits::evm_circuit::execution::coinbase::test::coinbase_gadget_test
             at ./src/evm_circuit/execution/coinbase.rs:109:9
   6: zkevm_circuits::evm_circuit::execution::coinbase::test::coinbase_gadget_test::{{closure}}
             at ./src/evm_circuit/execution/coinbase.rs:108:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/d914f17ca71a33a89b2dc3436fca51b1a091559e/library/core/src/ops/function.rs:227:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/d914f17ca71a33a89b2dc3436fca51b1a091559e/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test evm_circuit::execution::coinbase::test::coinbase_gadget_test ... FAILED

failures:

failures:
    evm_circuit::execution::coinbase::test::coinbase_gadget_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 55 filtered out; finished in 60.55s

error: test failed, to rerun pass '--lib'
```

It seemed to us that something is missing in regards the RLC and COINBASE in the PR that was made. It seems that the value that is introduced in the tables is with RLC in the `stack` and without in the `rw`. 

We suggest to take a look to: #287

Any thoughts @han0110 @DreamWuGit ?

**Comment by han0110** [comment]
It's because I am fixing the inconsistency between circuit and spec, but on the halfway. So I only updated the `table_assignments` of `BlockContext `, instead of updating the constraint of `CoinbaseGadget`.

https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/witness.rs#L45-L53

In spec, the `coinbase` exists as raw value in `block_table`, and we need to do decomposition then random linear combination to push it on stack. But i haven't fixed this part in circuit.

https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/execution/coinbase.rs#L32-L42

**Comment by ed255** [comment]
@han0110 this is how I resolved this issue in my PR for CALLER:

https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/execution/caller.rs#L31-L51

Do you think this approach is OK?

**Comment by han0110** [comment]
I will work in this way, but I think it would be even simpler to not query so many bytes for `caller_address`, because no such cases will happen, so we don't bother to check it.

https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/execution/caller.rs#L21-L24

```diff
-  caller_address: RandomLinearCombination<F, 32>,
+  caller_address: RandomLinearCombination<F, 20>,
```

**Comment by CPerezz** [comment]
Let's use the CONSTANT value for the `ADDRESS_BYTES`

**Comment by han0110** [comment]
It makes sense, actually there is already one here:

https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/param.rs#L16

**Comment by CPerezz** [comment]
> It makes sense, actually there is already one here:
> 
> https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/param.rs#L16

I remembered that there was one. But did not remember the name ahahhaha :rofl:

**Comment by icemelon** [comment]
cc @DreamWuGit 
I think we should not use RLC for COINBASE address in the block context table as the value is at most 20 bytes, which doesn't exceed the Fp field. Also could you enhance the test cases to cover COINBASE addresses other than 0x0?

**Comment by CPerezz** [comment]
> cc @DreamWuGit 
> I think we should not use RLC for COINBASE address in the block context table as the value is at most 20 bytes, which doesn't exceed the Fp field. Also could you enhance the test cases to cover COINBASE addresses other than 0x0?

There's already a lazy_static var defined with a non-zero value that you can use.

**Comment by DreamWuGit** [comment]
after taking look at PR #292 , just as Han said as before and this have been fixed in that pr, thanks!


> It's because I am fixing the inconsistency between circuit and spec, but on the halfway. So I only updated the `table_assignments` of `BlockContext`, but haven't updated the constraint of `CoinbaseGadget`.
> 
> https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/witness.rs#L45-L53
> 
> In spec, the `coinbase` exists as raw value in `block_table`, and we need to do decomposition then random linear combination to push it on stack. But i haven't fixed this part in circuit.
> 
> https://github.com/appliedzkp/zkevm-circuits/blob/7c79ae2edfaa249121a67933ceb2b45590e1d25c/zkevm-circuits/src/evm_circuit/execution/coinbase.rs#L32-L42

**Comment by CPerezz** [comment]
Closing them as was solved via #292 

Thanks @DreamWuGit and @ChihChengLiang for remembering me!
